### PR TITLE
Add Notification.close() to hide a notification from the server

### LINKED ISSUE #2114: API method to close Notification
**Originally by @johannesh2**

___
There should be an API method for closing a Notification shown with Window.showNotification(Notification). The method could be e.g. Window.hideNotification(Notification) or Notification.close().

This would be used in addition to the timeout. Example use case: If a timeout is 10 seconds for a notification, but the state of the application changes so that the notification is not relevant any more, it should be possible to hide the notification.

___
 Imported from https://dev.vaadin.com/ issue #8351

### PR DESCRIPTION
**Pull request by Artur-**
Also fixes the problem that notifications were never removed on the server side

Fixes #2114

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/10483)

### CODE REVIEW

File: uitest/src/test/java/com/vaadin/tests/elements/notification/NotificationCloseEventTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.vaadin.tests.elements.notification;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import com.vaadin.testbench.elements.ButtonElement;
+import com.vaadin.testbench.elements.NotificationElement;
+import com.vaadin.tests.tb3.SingleBrowserTest;
+
+public class NotificationCloseEventTest extends SingleBrowserTest {
```
Comment (tsuoanttila) [inline uitest/src/test/java/com/vaadin/tests/elements/notification/NotificationCloseEventTest.java]: Would it be possible to add a reloading test with `@PreserveOnRefresh` to the same UI and test class?

Comment (Artur-) [reply uitest/src/test/java/com/vaadin/tests/elements/notification/NotificationCloseEventTest.java]: Added

**Comment by tsuoanttila** [comment]
Other than a naming question, seems ok for me.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/ui/Notification.java, line 403 at r2](https://reviewable.io:443/reviews/vaadin/framework/10483#-L20dcFF4Rdoq5q_tAmx:-L20dcFF4Rdoq5q_tAmy:bjqycwl) ([raw file](https://github.com/vaadin/framework/blob/1a7676464c04429c263db3a16088e95427b76e7f/server/src/main/java/com/vaadin/ui/Notification.java#L403)):*
> ```Java
>      * @since
>      */
>     public void hide() {
> ```

Last remaining question from me is `hide` vs. `close`. What was the decision of the API naming based on?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483#-:-L20dsNwsewt5m7UkDvD:brz83or)*

**Comment by tsuoanttila** [comment]
Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483)*

**Comment by Artur-** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/ui/Notification.java, line 403 at r2](https://reviewable.io:443/reviews/vaadin/framework/10483#-L20dcFF4Rdoq5q_tAmx:-L20er4ZmEpJMGcW6lra:ban6ebh) ([raw file](https://github.com/vaadin/framework/blob/1a7676464c04429c263db3a16088e95427b76e7f/server/src/main/java/com/vaadin/ui/Notification.java#L403)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Last remaining question from me is `hide` vs. `close`. What was the decision of the API naming based on?
</blockquote></details>

Consistency with show. Could be ’close’ also to be consistent with Window, I do not have a strong opinion on this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483)*

**Comment by Artur-** [comment]
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/ui/Notification.java, line 403 at r2](https://reviewable.io:443/reviews/vaadin/framework/10483#-L20dcFF4Rdoq5q_tAmx:-L21dcA0iIlCpqwGz9kH:bq82mt0) ([raw file](https://github.com/vaadin/framework/blob/1a7676464c04429c263db3a16088e95427b76e7f/server/src/main/java/com/vaadin/ui/Notification.java#L403)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Consistency with show. Could be ’close’ also to be consistent with Window, I do not have a strong opinion on this
</blockquote></details>

Changed to `close`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483)*

**Comment by tsuoanttila** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/ui/Notification.java, line 408 at r3](https://reviewable.io:443/reviews/vaadin/framework/10483#-L24v7hc3oVkZb5-LHsF:-L24v7hc3oVkZb5-LHsG:bje5hyl) ([raw file](https://github.com/vaadin/framework/blob/5b9156e9d49eedd0f63e9354a50204a969b880b7/server/src/main/java/com/vaadin/ui/Notification.java#L408)):*
> ```Java
> 
>     /**
>      * Hides the notification.
> ```

Should the JavaDocs also speak of closing and not hiding ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483)*

**Comment by Artur-** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[server/src/main/java/com/vaadin/ui/Notification.java, line 408 at r3](https://reviewable.io:443/reviews/vaadin/framework/10483#-L24v7hc3oVkZb5-LHsF:-L266Y_J5Er720S_4pVj:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5b9156e9d49eedd0f63e9354a50204a969b880b7/server/src/main/java/com/vaadin/ui/Notification.java#L408)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Should the JavaDocs also speak of closing and not hiding ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483)*

**Comment by tsuoanttila** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10483)*

**Comment by wbadam** [comment]
Is the following code invalid? With removing the extension on close(), a notification cannot be shown for a second time.
```Java
sample = new Notification("Notification message", "Notification description");
showButton = new Button("Show notification", event -> sample.show(Page.getCurrent()));
```
Code example is from the Sampler and throws
```
java.lang.IllegalStateException: An extension can not be set to extend a new target after getting detached from the previous.
```

**Pull request merged**